Project aim - To get clear insights from New York TLC's data. Analyze data, identify key variables, ensure that dataset is ready for analysis.

Step 1 - Import Pandas and Numpy packages to work on the dataset

In [2]:
import pandas as pd

In [3]:
import numpy as np

Step 2 - Read the dataset by providing the location of the csv file

In [4]:
df = pd.read_csv(r'C:\Omkar\Omkar\Omkar Data Governance\Getting Started with Python - Google\Module 5\Automatidata scenario\2017_Yellow_Taxi_Trip_Data.csv')

View your dataset by the below code. You may change the count to understand the dataset better. I selected 20 rows.

In [5]:
df.head(20)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0     24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1     35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2    106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3     38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4     30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5     23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6     37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7     69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8      8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9     95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   
10    18017909         2  03/04/2017 11:58:00 AM  03/04/2017 12:13:12 PM   
11    18600059         2   03/05/2017 7:15:30 PM   03/05/2017 7:52:18 PM   
12    46782248         1   06/09/2017 7:00:26 PM   06/09/2017 7:20:11 PM   
13    94113247         2  11/06/2017 11:35:05 PM  11/06/2017 11:42:57 PM   
14    14168279         1   02/22/2017 3:18:31 PM   02/22/2017 3:42:50 PM   
15    47444401         2   06/02/2017 6:41:39 AM   06/02/2017 6:57:47 AM   
16    69088676         1   08/15/2017 7:48:08 PM   08/15/2017 8:00:37 PM   
17    58691513         2   07/10/2017 1:36:31 PM   07/10/2017 1:48:43 PM   
18    35388828         2   04/10/2017 6:12:58 PM   04/10/2017 6:17:39 PM   
19    18383214         2   03/05/2017 4:01:07 AM   03/05/2017 4:14:11 AM   

    passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                 6           3.34           1                  N   
1                 1           1.80           1                  N   
2                 1           1.00           1                  N   
3                 1           3.70           1                  N   
4                 1           4.37           1                  N   
5                 6           2.30           1                  N   
6                 1          12.83           1                  N   
7                 1           2.98           1                  N   
8                 1           1.20           1                  N   
9                 1           1.60           1                  N   
10                1           1.77           1                  N   
11                2          18.90           2                  N   
12                1           3.00           1                  N   
13                1           2.39           1                  N   
14                1           3.30           1                  N   
15                1           5.93           1                  N   
16                1           3.60           1                  N   
17                2           1.71           1                  N   
18                2           0.63           1                  N   
19                2           2.77           1                  N   

    PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0            100           231             1         13.0    0.0      0.5   
1            186            43             1         16.0    0.0      0.5   
2            262           236             1          6.5    0.0      0.5   
3            188            97             1         20.5    0.0      0.5   
4              4           112             2         16.5    0.5      0.5   
5            161           236             1          9.0    0.5      0.5   
6             79           241             1         47.5    1.0      0.5   
7            237           114             1         16.0    1.0      0.5   
8            234           249             2          9.0    0.0      0.5   
9            239           237             1         13.0    0.0      0.5   
10           162           142             1         11.5    0.0      0.5   
11           236           132   

Step 3 -
df.head()
df.info()
df.describe()
These three functions are used to understand the dataset.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

df.info() lets us understand that Dtypes are non-numeric. Two of which are datetime. No null values are found.

In [7]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

Using df.describe() we get fare amount distribution. The maximum fare amount is a much larger value than the 25-75 percent range of values or even the other ranges. There are negative values for fare amount. Regarding trip distance, most rides are between 1-3 miles and the maximum is over 33 miles.

Step 4 - The data needs to be further investigated and analyzed.
I sorted the data based on trip distance which which showed that the largest miles are from 25.86 to 33.96 miles for the top 20 counts.


In [10]:
df_sort = df.sort_values(by=['trip_distance'],ascending=False)

In [11]:
df_sort.head(20)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
9280     51810714         2  06/18/2017 11:33:25 PM  06/19/2017 12:12:38 AM   
13861    40523668         2   05/19/2017 8:20:21 AM   05/19/2017 9:20:30 AM   
6064     49894023         2  06/13/2017 12:30:22 PM   06/13/2017 1:37:51 PM   
10291    76319330         2  09/11/2017 11:41:04 AM  09/11/2017 12:18:58 PM   
29       94052446         2   11/06/2017 8:30:50 PM  11/07/2017 12:00:00 AM   
18130    90375786         1   10/26/2017 2:45:01 PM   10/26/2017 4:12:49 PM   
5792     68023798         2   08/11/2017 2:14:01 PM   08/11/2017 3:17:31 PM   
15350    77309977         2   09/14/2017 1:44:44 PM   09/14/2017 2:34:29 PM   
10302    43431843         1   05/15/2017 8:11:34 AM   05/15/2017 9:03:16 AM   
2592     51094874         2   06/16/2017 6:51:20 PM   06/16/2017 7:41:42 PM   
20612    67238347         2   08/08/2017 9:01:00 PM   08/08/2017 9:40:04 PM   
1908     81396251         2  09/27/2017 11:49:52 PM  09/28/2017 12:26:29 AM   
20545    71466504         1   08/24/2017 1:03:49 PM   08/24/2017 1:49:05 PM   
4138    100097208         2   11/26/2017 3:37:35 PM   11/26/2017 4:50:22 PM   
15169    71224174         2   08/23/2017 4:20:15 PM   08/23/2017 5:08:55 PM   
1496     27465882         2   04/03/2017 3:37:53 PM   04/03/2017 4:44:33 PM   
7217     13572254         1   02/20/2017 2:28:11 PM   02/20/2017 3:29:14 PM   
19483    34486843         2   04/26/2017 8:31:31 AM  04/26/2017 11:10:50 AM   
908      25075013         2   03/27/2017 1:01:38 PM   03/27/2017 1:38:44 PM   
4715     78418799         2   09/17/2017 8:04:24 PM   09/17/2017 9:02:24 PM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
9280                 2          33.96           5                  N   
13861                1          33.92           5                  N   
6064                 1          32.72           3                  N   
10291                1          31.95           4                  N   
29                   1          30.83           1                  N   
18130                1          30.50           1                  N   
5792                 1          30.33           2                  N   
15350                1          28.23           2                  N   
10302                1          28.20           2                  N   
2592                 1          27.97           2                  N   
20612                1          27.88           1                  N   
1908                 2          27.34           1                  N   
20545                1          27.20           1                  N   
4138                 2          26.86           1                  N   
15169                2          26.54           1                  N   
1496                 1          26.39           2                  N   
7217                 1          26.20           1                  N   
19483                1          26.12           2                  N   
908                  2          26.12           4                  N   
4715                 1          25.86           4                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
9280            132           265             2       150.00    0.0      0.0   
13861           229           265             1       200.01    0.0      0.5   
6064            138             1             1       107.00    0.0      0.0   
10291           138           265             2       131.00    0.0      0.5   
29              132            23             1        80.00    0.5      0.5   
18130           132           220             1        90.50    0.0      0.5   
5792            132           158             1        52.00    0.0      0.5   
15350            13           132             1        52.00    0.0      0.5   
10302            90           132             1        52.00    0.0      0.5   
2592            261           132 

Now the dataset is sorted based on the total amount.

In [44]:
total_amount_sorted = df.sort_values(
    ['total_amount'], ascending=False)['total_amount']
total_amount_sorted.head(20)

8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
1928      137.80
10291     131.80
6708      126.00
11608     123.30
908       121.56
7281      120.96
18130     119.31
13621     115.94
13359     111.95
29        111.38
Name: total_amount, dtype: float64

The top values for total_amount seems okay, now I checked the bottom 20 values.

In [14]:
total_amount_sorted.tail(20)

14283      0.31
19067      0.30
10506      0.00
4402       0.00
22566      0.00
5722       0.00
1646      -3.30
314       -3.80
18565     -3.80
5758      -3.80
10281     -4.30
5448      -4.30
4423      -4.30
8204      -4.80
20317     -4.80
11204     -5.30
14714     -5.30
20698     -5.80
17602     -5.80
12944   -120.30
Name: total_amount, dtype: float64

Bottom values are negative and total_amount cannot be negative. I manipulated the data by making those values positive.

In [51]:
total_amount_sorted_positive = df.sort_values(['total_amount'], ascending=False)['total_amount'].abs()
total_amount_sorted_positive.tail(20)

14283      0.31
19067      0.30
10506      0.00
4402       0.00
22566      0.00
5722       0.00
1646       3.30
314        3.80
18565      3.80
5758       3.80
10281      4.30
5448       4.30
4423       4.30
8204       4.80
20317      4.80
11204      5.30
14714      5.30
20698      5.80
17602      5.80
12944    120.30
Name: total_amount, dtype: float64

In [53]:
df['payment_type'].value_counts()

payment_type
1    15265
2     7267
3      121
4       46
Name: count, dtype: int64

According to the data dictionary, the payment method was encoded as follows:
1 = Credit card
2 = Cash
3 = No charge
4 = Dispute
5 = Unknown
6 = Voided trip

We have been asked -
How many of each payment type are represented in the data?
In our dataset we have four types.
1 = Credit card
2 = Cash
3 = No charge
4 = Dispute

In [54]:
avg_cc_tip = df[df['payment_type']==1]['tip_amount'].mean()
print('Avg. cc tip:', avg_cc_tip)

Avg. cc tip: 2.7298001965280054


The above code answers the next question -
# What is the average tip for trips paid for with credit card?
Avg. cc tip: 2.7298001965280054

In [55]:
avg_cash_tip = df[df['payment_type']==2]['tip_amount'].mean()
print('Avg. cash tip:', avg_cash_tip)

Avg. cash tip: 0.0


The above code answers the next question -
What is the average tip for trips paid for with cash?
Avg. cash tip: 0.0

In [58]:
#How many times is each vendor ID represented in the data?
df['VendorID'].value_counts()

VendorID
2    12626
1    10073
Name: count, dtype: int64

In [57]:
# What is the mean total amount for each vendor?
df.groupby(['VendorID']).mean(numeric_only=True)[['total_amount']]

total_amount
VendorID              
1            16.298119
2            16.320382

In [59]:
# Filter the data for credit card payments only
credit_card = df[df['payment_type']==1]

# Filter the credit-card-only data for passenger count only
credit_card['passenger_count'].value_counts()

passenger_count
1    10977
2     2168
5      775
3      600
6      451
4      267
0       27
Name: count, dtype: int64

In [60]:
# LAst ask is to calculate the average tip amount for each passenger count (credit card payments only)
credit_card.groupby(['passenger_count']).mean(numeric_only=True)[['tip_amount']]

tip_amount
passenger_count            
0                  2.610370
1                  2.714681
2                  2.829949
3                  2.726800
4                  2.607753
5                  2.762645
6                  2.643326